src/lib/tasks/run-hit-test.ts

```diff
@@ -0,0 +1,137 @@
+ */
+import REGL from 'regl';
+
+import {AttributeMapper} from '../attribute-mapper';
+import {HitTestParameters} from '../hit-test-types';
+import {SpriteImpl} from '../sprite-impl';
+import {InternalPropertiesSymbol} from '../symbols';
+
+/**
+ * To avoid circular imports, this file cannot depend on scene-internal.ts so
+ * here we define the minimum necessary API surface that the task implementation
+ * needs to operate.
+ */
+interface CoordinatorAPI {
+  instanceHitTestInputUvBuffer: REGL.Buffer;
+  instanceHitTestInputUvValues: Float32Array;
+  instanceHitTestInputIndexActiveBuffer: REGL.Buffer;
+  instanceHitTestInputIndexActiveValues: Float32Array;
+  instanceHitTestOutputUvBuffer: REGL.Buffer;
+  instanceSwatchUvValues: Float32Array;
+  hitTestAttributeMapper: AttributeMapper;
+  hitTestCommand?: () => void;
+  hitTestCount: number;
+  hitTestParameters: HitTestParameters;
+  hitTestOutputResults: Float32Array;
+  hitTestOutputValues: Uint8Array;
+  hitTestOutputValuesFramebuffer: REGL.Framebuffer2D;
+  regl: REGL.Regl;
+}
+
+/**
+ * Perform a hit test and read back the results.
+ *
+ * @param coordinator Upstream object upon which this task operates.
+ */
+export function runHitTest(
+    coordinator: CoordinatorAPI,
+) {
+  // These values are API-user provided, but are already be checked for
+  // correctness upstream in SceneInternal.
+  const {sprites, width, height, inclusive} = coordinator.hitTestParameters;
+
+  coordinator.hitTestCount = sprites.length;
+
+  const results =
+      coordinator.hitTestOutputResults.subarray(0, coordinator.hitTestCount);
+
+  // Short-circuit if the parameters guarantee there will be no hits.
+  if (!inclusive && (!width || !height)) {
+    console.warn('Inclusive hit test on a box with zero size always misses');
+    results.fill(-1);
+    return;
+  }
+
+  // Shorthand variables to make code more readable.
+  const inputUv = coordinator.instanceHitTestInputUvValues;
+  const indexActive = coordinator.instanceHitTestInputIndexActiveValues;
+  const swatchUv = coordinator.instanceSwatchUvValues;
+
+  // Copy swatch UVs into the input UV values array. This way, when the command
+  // runs, it will reference the correct swatches for the candidate sprites.
+  for (let i = 0; i < sprites.length; i++) {
+    const sprite = sprites[i] as SpriteImpl;
+    const swatchIndex = sprite[InternalPropertiesSymbol].index || 0;
+
+    inputUv[i * 2] = swatchUv[swatchIndex * 2];
+    inputUv[i * 2 + 1] = swatchUv[swatchIndex * 2 + 1];
+
+    indexActive[i * 2] = swatchIndex;
+    indexActive[i * 2 + 1] = sprite.isActive ? 1 : 0;
+  }
+
+  // Re-bind the UV and Index/Active values to their buffers.
+  coordinator.instanceHitTestInputUvBuffer(
+      inputUv.subarray(0, coordinator.hitTestCount * 2));
+  coordinator.instanceHitTestInputIndexActiveBuffer(
+      indexActive.subarray(0, coordinator.hitTestCount * 2));
+
+  // Invoke the hit test command.
+  coordinator.hitTestCommand!();
```

Comment:
This could be a personal preference, but I am not a huge fan of [non-null assertions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator) outside specific scenarios (e.g., querying elements in the DOM that you know will be there)... 

Is there a reason this is optional in the `CoordinatorAPI` but then assumed to exist here? Why not make this a required property of the `CoordinatorAPI`?

Reply:
Good point. Upgraded `hitTestCommand` to be a required property both here and in `scene-internal.ts`. Turns out it is always set during construction, so there was never a case where it would be missing anyway.